torch/csrc/stable/device_struct.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <torch/csrc/stable/c/shim.h>
+#include <torch/headeronly/core/DeviceType.h>
+#include <torch/headeronly/macros/Macros.h>
+#include <torch/headeronly/util/Exception.h>
+#include <torch/headeronly/util/shim_utils.h>
+
+#include <memory>
+
+HIDDEN_NAMESPACE_BEGIN(torch, stable)
+
+using DeviceType = torch::headeronly::DeviceType;
+
+// The torch::stable::Device class is a high-level C++ wrapper around
+// the C shim Device APIs. We've modeled this class after c10::Device
+// to provide a stable ABI-compatible device representation.
+//
+// This class provides:
+// 1. A nicer UX much closer to c10::Device than the C APIs with DeviceHandle
+// 2. Automatic memory management via shared_ptr
+// 3. Compatibility with the stable ABI guarantees
+class Device {
+ private:
+  std::shared_ptr<DeviceOpaque> dh_;
+
+ public:
+  // Construct a stable::Device from a DeviceType and optional device index
+  // Default index is -1 (current device)
+  /* implicit */ Device(DeviceType type, int32_t index = -1) {
+    DeviceHandle ret;
+    TORCH_ERROR_CODE_CHECK(
+        torch_create_device(static_cast<int32_t>(type), index, &ret));
```

Comment:
I suspect we need to call `from` and `to` on the type here instead of static casting -- and everywhere else we input the type into the shim